### PR TITLE
feat(explore): Show confirmation modal if user exits Explore without saving changes

### PR DESCRIPTION
**Pull request by kgabryje**
### SUMMARY
This PR adds a browser confirmation modal (similar to the one that we have in dashboard edit mode) when user makes a change in control panel and then tries to close Explore (either by closing the tab or moving to another page) without clicking Save button.
The modal that displays when user navigates out of SPA context (some external page or a Superset page that hasn't been migrated to SPA yet) or closes the tab, is not customizable.
In the case of navigating within SPA context, we can easily display a custom message in the modal thanks to react-router's `Prompt` component. However, displaying a custom modal instead of the default one rendered by `<Prompt />` requires more effort and will be handled in a separate PR.

CC @kasiazjc 

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

https://user-images.githubusercontent.com/15073128/188621562-23b4e1cf-32e7-440a-9b03-ae90cdf398d4.mov

### TESTING INSTRUCTIONS
1. Open a chart and make some changes in control panel
2. Try to go to another SPA page (like charts list)
3. A modal should appear. Clicking Leave should discard your changes and go to next page. Clicking Cancel should close the modal.
4. Make some changes in control panel and try to close browser tab or go to some external page.
5. A default browser confirmation modal should appear (it's not customizable)
6. Save your changes
7. Now leaving the page should not trigger the confirmation modal

KNOWN LIMITATION: if the previous page is a SPA route (for example you opened Explore from charts list), clicking browser back button will not trigger the confirmation modal. It's very bugged in react-router and I decided to disable it for now. We'll try to find a workaround in the future

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by codyml** [comment]
@kgabryje I'm not sure if this is fixable or if it was like this before, but I noticed that if you have unsaved changes and try to navigate to Dashboard, even if you click "cancel" the navigation indicator stays on Dashboard:
Screen Shot 2022-07-29 at 10 58 12 AM

**Comment by codyml** [comment]
@kgabryje Also not sure if this was intentional, but I noticed that if you edit the chart title but clicking directly on the rendered title and edit it, the "Altered" tag appears but it doesn't block navigation away

**Comment by kgabryje** [comment]
> @kgabryje Also not sure if this was intentional, but I noticed that if you edit the chart title but clicking directly on the rendered title and edit it, the "Altered" tag appears but it doesn't block navigation away.

That's an existing bug - you can try it by going to charts list, then, dashboards list, then click "back" - dashboards will still be highlighted

**Comment by kgabryje** [comment]
> @kgabryje I'm not sure if this is fixable or if it was like this before, but I noticed that if you have unsaved changes and try to navigate to Dashboard, even if you click "cancel" the navigation indicator stays on Dashboard

That's an existing bug - you can try it by going to charts list, then, dashboards list, then click "back" - dashboards will still be highlighted

**Comment by kgabryje** [comment]
> @kgabryje Also not sure if this was intentional, but I noticed that if you edit the chart title but clicking directly on the rendered title and edit it, the "Altered" tag appears but it doesn't block navigation away.

Oops, well spotted!

File: superset-frontend/src/components/RouteLeaveGuard/RouteLeaveGuard.tsx [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import React, { useEffect, useRef } from 'react';
+import { Prompt } from 'react-router-dom';
+// eslint-disable-next-line import/no-extraneous-dependencies
+import { PromptProps } from 'react-router';
+
+const handleUnloadEvent = (e: BeforeUnloadEvent) => {
```
[the remark this answers is not in the document]
Comment (kgabryje) [reply superset-frontend/src/components/RouteLeaveGuard/RouteLeaveGuard.tsx:25]: Good point!

**Comment by eschutho** [comment]
@jinghua-qa can we test what happens when the chart is running on a query? Does it show the correct save chart modal with the dialog to save a dataset?

**Comment by jinghua-qa** [comment]
in the confirmation modal, is it make sense to have a cancel option or close dialog option? Looks like now only have discard and save, as user i may need to think about double check my change before i save, but i dont have an option currenlty.
Screen Shot 2022-08-01 at 12 29 50 AM

**Comment by jinghua-qa** [comment]
I also noticed that there are different confirmation dialog, and this one did not have save option, only cancel and leave, is this expected? are we trying to have same confirmation dialog modal when user leave explore without save?
Screen Shot 2022-08-01 at 12 33 52 AM

**Comment by kgabryje** [comment]
> in the confirmation modal, is it make sense to have a cancel option or close dialog option? Looks like now only have discard and save, as user i may need to think about double check my change before i save, but i dont have an option currenlty. Screen Shot 2022-08-01 at 12 29 50 AM

I think it makes sense! @kasiazjc wdyt?

**Comment by kgabryje** [comment]
> I also noticed that there are different confirmation dialog, and this one did not have save option, only cancel and leave, is this expected? are we trying to have same confirmation dialog modal when user leave explore without save? Screen Shot 2022-08-01 at 12 33 52 AM

Yes - our custom confirmation modal is possible to display only when navgating between SPA routes. When you try to close the tab or go to an external page, a default browser modal appears - it's not customizable in any way (it's a browser security measure)

**Comment by kasiazjc** [comment]
> > in the confirmation modal, is it make sense to have a cancel option or close dialog option? Looks like now only have discard and save, as user i may need to think about double check my change before i save, but i dont have an option currenlty. Screen Shot 2022-08-01 at 12 29 50 AM
> 
> I think it makes sense! @kasiazjc wdyt?

Good point, I'm thinking - maybe for consistency we should follow chrome flow actually instead of adding this one. 
Copy: 

Your chart is not saved
If you leave and don't save, changes will be lost. 
[Cancel][Leave]

Thoughts?

**Comment by jinghua-qa** [comment]
> > I think it makes sense! @kasiazjc wdyt?
> 
> Good point, I'm thinking - maybe for consistency we should follow chrome flow actually instead of adding this one. Copy:
> 
> Your chart is not saved If you leave and don't save, changes will be lost. [Cancel][Leave]
> 
> Thoughts?

Do we have a final plan for this design? My personal feeling is will be great to have all, [Cancel][Leave][Save], if that is too complicated, i think i will vote for [Cancel][Leave] which is consistent with the chrome flow, and i can always cancel and do save on explore.

**Comment by kasiazjc** [comment]
> > > I think it makes sense! @kasiazjc wdyt?
> > 
> > 
> > Good point, I'm thinking - maybe for consistency we should follow chrome flow actually instead of adding this one. Copy:
> > Your chart is not saved If you leave and don't save, changes will be lost. [Cancel][Leave]
> > Thoughts?
> 
> Do we have a final plan for this design? My personal feeling is will be great to have all, [Cancel][Leave][Save], if that is too complicated, i think i will vote for [Cancel][Leave] which is consistent with the chrome flow, and i can always cancel and do save on explore.

I try to avoid adding 3 buttons as much as possible, as it can get quite confusing. After clicking "save" we would still have to open "save" modal, so I would say let's go with [Cancel][Leave] @kgabryje

**Comment by codyml** [comment]
> > > > I think it makes sense! @kasiazjc wdyt?
> > > 
> > > 
> > > Good point, I'm thinking - maybe for consistency we should follow chrome flow actually instead of adding this one. Copy:
> > > Your chart is not saved If you leave and don't save, changes will be lost. [Cancel][Leave]
> > > Thoughts?
> > 
> > 
> > Do we have a final plan for this design? My personal feeling is will be great to have all, [Cancel][Leave][Save], if that is too complicated, i think i will vote for [Cancel][Leave] which is consistent with the chrome flow, and i can always cancel and do save on explore.
> 
> I try to avoid adding 3 buttons as much as possible, as it can get quite confusing. After clicking "save" we would still have to open "save" modal, so I would say let's go with [Cancel][Leave] @kgabryje

[Cancel] [Leave] makes the most sense to me too!  As a user I think I'd expect there to be an option to just cancel the close.

**Comment by kgabryje** [comment]
Updated with Cancel/Leave

image

**Comment by kasiazjc** [comment]
> Updated with Cancel/Leave
> 
> image

lgtm, thankss

File: superset-frontend/src/explore/components/ExploreViewContainer/index.jsx [hunk 2]
```diff
@@ -337,7 +375,7 @@ function ExploreViewContainer(props) {
   }
 
   function toggleModal() {
-    setShowingModal(!showingModal);
+    props.actions.setSaveModalVisible(!props.isSaveModalVisible);
```
Comment (codyml) [inline superset-frontend/src/explore/components/ExploreViewContainer/index.jsx:378]: I think this PR might be re-triggering the annoying bug that forced us to [revert these two PRs](https://github.com/apache/superset/pull/20796):

https://user-images.githubusercontent.com/13007381/188811425-a306f8b2-4ec9-4aa1-94f7-504759ea1a6f.mov

Comment (codyml) [reply superset-frontend/src/explore/components/ExploreViewContainer/index.jsx:378]: I don't think we were able to solve it: IIRC it seemed to happen whenever we tried to send additional Redux actions at the same time as when Explore was rendering.  [Here's the PR](https://github.com/apache/superset/pull/20704) that would have fixed it if we hadn't reverted the whole thing.  Not sure if your case is similar?  I think that also might be why Cypress is failing on yours.

Comment (codyml) [reply superset-frontend/src/explore/components/ExploreViewContainer/index.jsx:378]: I also noticed that if I Save As and it doesn't refresh, it'll then show the modal every time even if I save it again.

**Pull request closed without merging**